### DISCUSSION TITLE
Does AroCC support parsing a `[]u8` in some way?

### DISCUSSION THREAD

**Discussion by heysokam** [Vexu/arocc#796] [Q&A] [answered]
Title, essentially.

Lets say I have a source code slice in memory.
How would I go about setting up what the parser requires to do its job?

I saw `Parser.parse(PP)`, but it requires a preprocessor object that in turn requires a `Compilation` which takes a folder... :thinking:
Am I missing something? How would one go about doing this?

**Comment by Vexu** [comment, marked as answer]
`Compilation` takes a folder that should be used as cwd, it should probably use a `Driver.Filesystem` to provide even better encapsulation with `Filesystem.fake`. Easiest way to get a `Compilation` is to just pass it `std.fs.cwd()`. Then you can use `Compilation.addSourceFromBuffer()` to turn your `[]u8` into a `Souce` that can be passed to `Preprocessor.preprocessSources()`.

The minimal steps to turn a source buffer into an AST would look something like this:
```zig
var comp = aro.Compilation.initDefault(gpa, std.fs.cwd());
defer comp.deinit();

const your_source = try comp.addSourceFromBuffer("<your buffer>", your_buffer);
const builtin = try comp.generateBuiltinMacros(.include_system_defines, null);

var pp = try aro.Preprocessor.initDefault(&comp);
defer pp.deinit();

try pp.preprocessSources(&.{ source, builtin });

var tree = try pp.parse();
defer tree.deinit();
```

**Reply by heysokam** [reply]
Thats so useful! Ty!
